packages/netstring/reader.js

```diff
@@ -2,67 +2,113 @@
+  //   - A newly allocated buffer large enough to accommodate the whole expected data.
+  //   In either case, remainingDataLength contains the length of the data to read.
+  //   If the whole data is received in one chunk, no copy is made.
+  /** @type {number[] | null} */
+  let lengthBuffer = [];
+  /** @type {Uint8Array | null} */
+  let dataBuffer = null;
+  let remainingDataLength = -1;
+
   for await (const chunk of input) {
-    if (length + chunk.byteLength >= capacity) {
-      while (length + chunk.byteLength >= capacity) {
-        capacity *= 2;
-      }
-      const replacement = new Uint8Array(capacity);
-      replacement.set(buffer, 0);
-      buffer = replacement;
-    }
-    buffer.set(chunk, length);
-    length += chunk.byteLength;
+    let buffer = chunk;
 
-    let drained = false;
-    while (!drained && length > 0) {
-      const colon = buffer.indexOf(COLON);
-      if (colon === 0) {
-        throw new Error(
-          `Expected number before colon at offset ${offset} of ${name}`,
-        );
-      } else if (colon > 0) {
-        const prefixBytes = buffer.subarray(0, colon);
-        const prefixString = decoder.decode(prefixBytes);
-        const contentLength = +prefixString;
-        if (Number.isNaN(contentLength)) {
-          throw new Error(
-            `Invalid netstring prefix length ${prefixString} at offset ${offset} of ${name}`,
-          );
+    while (buffer.length) {
+      // Waiting for full length prefix
+      if (lengthBuffer) {
+        let i = 0;
+        while (i < buffer.length) {
+          const c = buffer[i];
+          i += 1;
+          if (c >= ZERO && c <= NINE) {
+            lengthBuffer.push(c);
+          } else if (c === COLON && lengthBuffer.length) {
+            lengthBuffer.push(c);
+            break;
+          } else {
+            throw new Error(
+              `Invalid netstring length prefix ${JSON.stringify(
+                String.fromCharCode(...lengthBuffer, c),
+              )} at offset ${offset} of ${name}`,
+            );
+          }
         }
-        const messageLength = colon + contentLength + 2;
-        if (messageLength <= length) {
-          yield buffer.subarray(colon + 1, colon + 1 + contentLength);
-          buffer.copyWithin(0, messageLength);
-          length -= messageLength;
-          offset += messageLength;
-        } else {
-          drained = true;
+
+        buffer = buffer.subarray(i);
+
+        if (lengthBuffer[lengthBuffer.length - 1] === COLON) {
+          lengthBuffer.pop();
+          const prefix = String.fromCharCode(...lengthBuffer);
+          remainingDataLength = +prefix;
+          if (Number.isNaN(remainingDataLength)) {
+            throw new Error(
+              `Invalid netstring prefix length ${prefix} at offset ${offset} of ${name}`,
+            );
+          }
+          offset += lengthBuffer.length + 1;
+          lengthBuffer = null;
+        }
+      }
+
```

Comment:
I unfortunately need to lean heavily on tests to verify the logic in here. I don’t see any flaws but from experience writing similar code, I’m sure that’s no indication. A fuzzer would make me more comfortable. A boundary searcher would be even better. A boundary searcher would be a test that visits various permutations of message sizes +/-1 around powers of two or other likely boundary cases, in various orders, trying to tease out edge cases.

Reply:
I'm not sure what powers of 2 have to do with the length prefix logic here. I believe the unit tests already cover most edge cases, which are:
- break in the middle of the prefix
- break between the number prefix and the colon
- invalid character in the prefix

The one case not covered is an empty prefix before the colon, and I can definitely fix that.

Can you think of any other cases to test for?

Reply:
Added an empty prefix test